omniscidb/L0Mgr/L0Mgr.cpp

```diff
@@ -139,8 +139,87 @@ void* allocate_device_mem(const size_t num_bytes, L0Device& device) {
   return mem;
 }
 
-L0Device::L0Device(const L0Driver& driver, ze_device_handle_t device)
+L0DataFetcher::L0DataFetcher(const L0Driver& driver, ze_device_handle_t device)
     : device_(device), driver_(driver) {
+  ze_command_queue_desc_t command_queue_fetch_desc = {
+      ZE_STRUCTURE_TYPE_COMMAND_QUEUE_DESC,
+      nullptr,
+      0,
+      0,
+      0,
+      ZE_COMMAND_QUEUE_MODE_ASYNCHRONOUS,
+      ZE_COMMAND_QUEUE_PRIORITY_NORMAL};
+  L0_SAFE_CALL(zeCommandQueueCreate(
+      driver.ctx(), device_, &command_queue_fetch_desc, &queue_handle_));
+  current_cl_bytes = {{}, 0};
+  L0_SAFE_CALL(
+      zeCommandListCreate(driver.ctx(), device_, &cl_desc, &current_cl_bytes.first));
+}
+
+L0DataFetcher::~L0DataFetcher() {
+  zeCommandQueueDestroy(queue_handle_);
+  zeCommandListDestroy(current_cl_bytes.first);
+  for (auto& dead_handle : graveyard) {
+    zeCommandListDestroy(dead_handle);
+  }
+  for (auto& cl_handle : recycled) {
+    zeCommandListDestroy(cl_handle);
+  }
+}
+
+void L0DataFetcher::recycleGraveyard() {
+  while (recycled.size() < GRAVEYARD_LIMIT && graveyard.size()) {
+    recycled.push_back(graveyard.front());
+    graveyard.pop_front();
+    L0_SAFE_CALL(zeCommandListReset(recycled.back()));
+  }
+  for (auto& dead_handle : graveyard) {
+    L0_SAFE_CALL(zeCommandListDestroy(recycled.back()));
+  }
+  graveyard.clear();
+}
+
+void L0DataFetcher::appendCopyCommand(void* dst,
+                                      const void* src,
+                                      const size_t num_bytes) {
+  std::unique_lock<std::mutex> cl_lock(current_cl_lock);
+  L0_SAFE_CALL(zeCommandListAppendMemoryCopy(
+      current_cl_bytes.first, dst, src, num_bytes, nullptr, 0, nullptr));
+  current_cl_bytes.second += num_bytes;
+  if (current_cl_bytes.second >= 128 * 1024 * 1024) {
+    ze_command_list_handle_t cl_h_copy = current_cl_bytes.first;
+    graveyard.push_back(current_cl_bytes.first);
+    current_cl_bytes = {{}, 0};
+    if (recycled.size()) {
+      current_cl_bytes.first = recycled.front();
+      recycled.pop_front();
+    } else {
+      L0_SAFE_CALL(
+          zeCommandListCreate(driver_.ctx(), device_, &cl_desc, &current_cl_bytes.first));
+    }
```

Comment:
This logic could be wrapped into smth like `getRecycledOrNew...()` for readability.